### PR TITLE
[OpenMP] Reorganize the initialization of `PluginAdaptorTy`

### PR DESCRIPTION
**Pull request by jdoerfert**
This introduces checked errors into the creation and initialization of `PluginAdaptorTy`. We also allow the adaptor to "hide" devices from the user if the initialization failed. The new organization avoids the "initOnce" stuff but we still do not eagerly initialize the plugin devices (I think we should merge `PluginAdaptorTy::initDevices` into `PluginAdaptorTy::init`)

### CODE REVIEW

File: openmp/libomptarget/src/PluginManager.cpp [hunk 1]
```diff
@@ -92,45 +112,61 @@ void PluginManager::init() {
...
 
   // Initialize the index of this RTL and save it in the used RTLs.
-  Plugin.DeviceOffset = Start;
+  DeviceOffset = ExclusiveDevicesAccessor->size();
 
   // If possible, set the device identifier offset in the plugin.
-  if (Plugin.set_device_offset)
-    Plugin.set_device_offset(Start);
+  if (set_device_offset)
+    set_device_offset(DeviceOffset);
```
Comment (shiltian) [inline openmp/libomptarget/src/PluginManager.cpp:145]: Why was this function named this way? This function is not like some others that are actually function pointers of C library/plugin functions. Of course it is not this patch's issue.

**Review by shiltian — APPROVED** [review]
LG with one nit.

File: openmp/libomptarget/src/PluginManager.cpp (hunk 1 above)
Comment (jdoerfert) [reply openmp/libomptarget/src/PluginManager.cpp:145]: It is like the others. I moved it into the PluginAdaptorTy so it checks the function pointer and calls it, like other plugin functions.

**Comment by dhruvachak** [comment]
After this reorganization, all available devices are getting initialized eagerly through  PluginAdaptorTy::initDevices. Before this patch, the device initialization would occur on demand through __tgt_target_kernel. While I don't know whether this leads to any measurable performance degradation, quite a bit of unnecessary work is being done now. For example, on an 8-GPU system, we are initializing all 8 devices now whereas previously we would initialize only 1 device (assuming that's all that was used by the program). Any reason why all devices are initialized eagerly?

Additionally, this could have implications for overheads incurred when OMPT-based tools are involved. According to the spec, the runtime shall invoke the device_initialize callback defined by the tool when a device is initialized. So after this patch, this callback will be invoked for all devices which in turn could do quite a bit of unnecessary metadata setup, etc.

I noticed another change because of this reorganization. A device is now getting initialized before main is called whereas previously it would be called from main through __tgt_target_kernel. This implies that the device initialization callback will now be called before main. While the spec does not say when this callback should be invoked, this change could cause problems for tools if static data structures were used by the tool. Perhaps this change is ok since the spec does not mandate any order. But again, is there a reason why this change was made?

@jdoerfert

**Comment by doru1004** [comment]
Would it be possible to have this behavior by default but also have the old behaviour optional?

File: openmp/libomptarget/plugins-nextgen/common/src/PluginInterface.cpp [hunk 2]
```diff
@@ -1704,7 +1704,7 @@ int32_t __tgt_rtl_number_of_devices() { return Plugin::get().getNumDevices(); }
 
 int64_t __tgt_rtl_init_requires(int64_t RequiresFlags) {
   Plugin::get().setRequiresFlag(RequiresFlags);
-  return RequiresFlags;
```
Comment (doru1004) [inline openmp/libomptarget/plugins-nextgen/common/src/PluginInterface.cpp:1707]: Why has this changed? It looks unrelated.

**Comment by Thyre** [comment]
> After this reorganization, all available devices are getting initialized eagerly through PluginAdaptorTy::initDevices. Before this patch, the device initialization would occur on demand through __tgt_target_kernel. While I don't know whether this leads to any measurable performance degradation, quite a bit of unnecessary work is being done now. For example, on an 8-GPU system, we are initializing all 8 devices now whereas previously we would initialize only 1 device (assuming that's all that was used by the program). Any reason why all devices are initialized eagerly?
> 
> Additionally, this could have implications for overheads incurred when OMPT-based tools are involved. According to the spec, the runtime shall invoke the device_initialize callback defined by the tool when a device is initialized. So after this patch, this callback will be invoked for all devices which in turn could do quite a bit of unnecessary metadata setup, etc.
> 
> I noticed another change because of this reorganization. A device is now getting initialized before main is called whereas previously it would be called from main through __tgt_target_kernel. This implies that the device initialization callback will now be called before main. While the spec does not say when this callback should be invoked, this change could cause problems for tools if static data structures were used by the tool. Perhaps this change is ok since the spec does not mandate any order. But again, is there a reason why this change was made?
> 
> @jdoerfert

I guess having all devices initialized early might help with https://github.com/ROCm/ROCm/issues/2057. However, we moved away from using the returned value a while ago. So this doesn't matter for us. 

As for potential additional overhead incurred with OMPT tools, I'll test how this MR affects Score-P. Looking at our code, we will initialize the device for measurements including locations (if tracing is available). Therefore, users will see empty locations for devices not being used. This isn't the case with other accelerator adapters. 

Especially the initialization before main worries me a bit, since #69318 occurs when one tries to do certain initialization during OMPT init when offloading is used.
